📖 Story Interactive docs (#30411)
ampproject/amphtml: extensions/amp-story-interactive/amp-story-interactive.md

<div layout="container" width="3" height="2">
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/quiz-art.png" layout="responsive" width="200" height="350"/></div>
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/animal-poll.png" layout="responsive" width="200" height="350"/></div>
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/results-art.png" layout="responsive" width="200" height="350"/></div>
</div>

## Usage

The amp-story-interactive component encompasses a set of interactive experiences. Specify an interactive experience by defining one of the elements below. For best results, only use one element per amp-story-page.

If you want to see all the components in action, check out the [example story](https://amp.dev/documentation/examples/components/amp-story-interactive-poll/story#page=title-components)

### amp-story-interactive-binary-poll

> The `amp-story-interactive-binary-poll` element provides a two option voting user interface. Users may select one of two valid options. When selected, the highlighted option fills the container and displays the total percentage of votes.

Reviewer: Can we add something to tell where the total percentage of votes is coming from? And that the user answer is persisted on page reload or new visit?

Author: Sounds good. Adding: "Most elements require a backend endpoint that will store aggregate data for each interactive, as well as persist the selected option for a user across sessions." to the Usage header

Does not support pairing with `amp-story-interactive-results`, and can optionally have a prompt.

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/binary-poll-raw.png" layout="intrinsic" width="400" height="230">

```html
<amp-story-interactive-binary-poll
  id="pizza-binary-poll"

>   endpoint="https://endpoint.com/v1/interactives"

Reviewer: This needs to ship with the Google backend as the default endpoint

Author: Can't include it here yet since it's not available. When it becomes available I plan to update these examples. Notice the "coming soon" on the backend section 😁

Reviewer: I think we should wait until it's ready before submitting this documentation!

Author: Independently, having the docs approved is a good way of getting ready for launching!

Author: It supports any backends that conform to the API described in the "Aggregate data source" section. Anyone can make a backend with that API and plug the endpoint here. Google will provide an endpoint for convenience, so we will add the endpoint to the docs once we have everything ready.

  prompt-text="Like Pizza?"

>   option-1-text="Yes"

Reviewer: Would be good to know that endpoint/text are mandatory, and that prompt, confetti are optional

Author: Isn't that what the Attributes section is for? There we have what attributes are mandatory and what are optional

  option-1-confetti="🍕"
  option-2-text="No"
  option-2-confetti="🤢"
>
</amp-story-interactive-binary-poll>
```

### amp-story-interactive-poll

The `amp-story-interactive-poll` element provides a voting experience with 2-4 options displayed vertically, where all options are valid. When selected, each option displays the total percentage of votes.

Display different categories based on user poll answers by pairing `amp-story-interactive-poll` with `amp-story-interactive-results`. Add a prompt for extra context.

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/poll-raw.png" layout="intrinsic" width="400" height="450">

[sourcecode:html]
<amp-story-interactive-poll
    id="season-poll"
    theme="dark"
    endpoint="https://endpoint.com/v1/interactives"
    prompt-text="Pick a season"
    option-1-text="Spring" option-1-confetti="🌼"
    option-2-text="Summer" option-2-confetti="☀️"
    option-3-text="Fall" option-3-confetti="🍁"
    option-4-text="Winter" option-4-confetti="☃️">
</amp-story-interactive-poll>
[/sourcecode]

### amp-story-interactive-quiz

> The `amp-story-interactive-quiz` element provides a guessing experience with 2-4 options, one of which is correct. It displays the voting percentages after the user makes a selection. The user selection is green if correct and red if incorrect.

Reviewer: > the voting percentages

Might be hard to understand without context. Can we do something that specifically mentions that it shows what other users answered as %?

Author: Will make a better effort at explaining this above all the descriptions for each element here. It's an important part of them

Display different categories based on percentage of correct user answers by pairing `amp-story-interactive-quiz` with `amp-story-interactive-results`. Add a prompt for extra context.

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/quiz-raw.png" layout="intrinsic" width="400" height="450">

[sourcecode:html]
<amp-story-interactive-quiz
    id="arts-quiz"
    endpoint="https://endpoint.com/v1/interactives"
    prompt-text='Who was the artist that created the famous painting "The Last Supper"?'
    option-1-text="Michelangelo"
    option-2-text="Leonardo da Vinci" option-2-correct option-2-confetti="🎨"
    option-3-text="Rahael"
    option-4-text="Sandro Boticelli">
</amp-story-interactive-quiz>
[/sourcecode]

### amp-story-interactive-results

> The `amp-story-interactive-result` element displays a customized state defined by the user's selection from previous polls or quizzes. This element requires use of a poll or quiz from previous pages to calculate an answer-based state. Each result state uses a category that may include an image, title and description to display to the user.

Reviewer: Publishers have to use exclusively polls or quizzes for it to work, right? Can we mention that as well?

Author: Yes! Currently we have "This element requires use of a poll or quiz from previous pages to calculate an answer-based state.". Anything else we are missing?

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/results-raw.png" layout="intrinsic" width="400" height="500">

[sourcecode:html]
<amp-story-interactive-results
    theme="dark"
    prompt-text="You are a"
    option-1-text="Dog" option-1-image="dog.png"
    option-1-description="You always have energy and love being with friends. Outdoors is your favorite place"
    option-2-text="Cat" option-2-image="cat.png"
    option-2-description="Cats are great animals for WFH">
</amp-story-interactive-results>
[/sourcecode]

## Attributes

The interactive experience elements from `amp-story-interactive` share an API language for customizing options.

### id (required for binary-poll, poll, quiz)

Element ID that identifies the interactive component in the story, and is used to compose the [`interactiveId`](#interactiveId) that is sent to the backend. Should be unique for each component in the story.

### endpoint (required for binary-poll, poll, quiz)

> Complete URL of backend. Stores interactive element votes.

Reviewer: Let's add something about the default Google hosted backend?

Author: Even though we already have the other section on the backends?

Author: Google will have a backend that conforms to the API described in `Aggregate data source` that anyone will be able to plug and use right away

### Themes

The `theme` attribute controls the chip color and text color of the component. Can be `light` (default) or `dark`.
The `chip-style` attribute controls the style details of the component. Can be `flat` (default), `shadow` or `transparent`.

## Sizing

The component follows the [container](https://amp.dev/documentation/guides-and-tutorials/learn/amp-html-layout/layouts_demonstrated/#container) model. The size can be changed by overriding the element's font-size, which by default will be set to `3*var(--story-page-vmin)`. The component has a `min-width: 14em` and `max-width: 25em`, unless it's a binary-poll or results, in which case it will have `max-width:18em`. This makes it occupy 75% of the width on portrait stories by default, but the width can be overridden with a CSS rule to any value in between the min and max. The height will depend on the font-size and number of lines on the prompt, so it cannot be specified.

### Creating pixel-perfect layouts using amp-story-grid-layer[aspect-ratio]

> While the component by default adapts to the screen size with the variable font-size, it doesn't stay perfectly consistent across screen sizes. It's possible to use the aspect-ratio layer in order to create layouts that will scale perfectly with different screen sizes, by setting the font-size in ems on the component.

Reviewer: Let's sync on this one offline, I think you wrote this with a ton of context in mind that very very few people have. We should think of how to better communicate all this context.

Author: Sure! I rewrote a bit after following the suggestions in previous comments from @CrystalOnScript but will show you

---
$category@: dynamic-content
formats:
  - stories
teaser:
  text: A rich set of interactive experiences for stories, including quizzes, polls and results.
draft:
  - true
tags:
  - dynamic-content
  - social
author: mszylkowski
---

<!---
Copyright 2020 The AMP HTML Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS-IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
-->

> [tip type="important"]Change image sources to ampproject/amphtml/master!!![/tip]

Author: This is a note to change this before submitting

> The amp-story-interactive component provides a set of experiences, such as quizzes or polls, for users in [Web stories](https://amp.dev/documentation/guides-and-tutorials/start/create_successful_stories/?format=stories). Interactive experiences provided by amp-story-interactive can integrate into tools or editors and support analytics.

Author: Hmmm "interactive experiences" is plural, why would it be "supports"? I can rewrite this part if that's confusing

<div layout="container" width="3" height="2">
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/quiz-art.png" layout="responsive" width="200" height="350"/></div>
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/animal-poll.png" layout="responsive" width="200" height="350"/></div>
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/results-art.png" layout="responsive" width="200" height="350"/></div>
</div>

## Interactive experiences

> The amp-story-interactive extension encompasses a set of interactive experiences. Specify an interactive experience by defining one of the elements below. For best results, only use one element per amp-story-page.

Author: Didn't seem to find many other places where I used "extension", but updated them (eg: in examples and guides)

If you want to see all the components in action, check out the [example story](/documentation/examples/components/amp-story-interactive-poll/story#page=title-components)

### amp-story-interactive-binary-poll

The amp-story-interactive-binary-poll element provides a two option voting user interface. Users may select one of two valid options. When selected, the highlighted option fills the container and displays the total percentage of votes.

> Is well suited to be used without amp-story-interactive-results, and can optionally have a prompt.

Author: Updated

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/binary-poll-raw.png" layout="intrinsic" width="400" height="230">

```html
<amp-story-interactive-binary-poll
  id="pizza-binary-poll"
  endpoint="https://endpoint.com/v1/interactives"
  prompt-text="Like Pizza?"
  option-1-text="Yes"
  option-1-confetti="🍕"
  option-2-text="No"
  option-2-confetti="🤢"
>
</amp-story-interactive-binary-poll>
```

### amp-story-interactive-poll

The amp-story-interactive-poll element provides a voting experience with 2-4 options displayed vertically, where all options are valid. When selected, each option displays the total percentage of votes.

> Can be paired up with amp-story-interactive-results to display different categories based on the answers to polls, but can be used independently as well. It is encouraged to add a prompt for extra context.

Author: Changed for "Add a poll for extra content" so we don't say we encourage it, even though most people will likely use it. The way to add a prompt is to use `prompt-text` as it can be seen in the attributes section. The examples also have the prompt-text attribute

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/poll-raw.png" layout="intrinsic" width="400" height="450">

[sourcecode:html]
<amp-story-interactive-poll
    id="season-poll"
    theme="dark"
    endpoint="https://endpoint.com/v1/interactives"
    prompt-text="Pick a season"
    option-1-text="Spring" option-1-confetti="🌼"
    option-2-text="Summer" option-2-confetti="☀️"
    option-3-text="Fall" option-3-confetti="🍁"
    option-4-text="Winter" option-4-confetti="☃️">
</amp-story-interactive-poll>
[/sourcecode]

### amp-story-interactive-quiz

The amp-story-interactive-quiz element provides a guessing experience with 2-4 options, one of which is correct. The option selected gets highlighted with green if correct, and red if not; and the percentages are displayed on the options according to the percentages of votes.

> Can be paired up with amp-story-interactive-results to display a score based on the correctness of the answers to quizzes, but can be used independently as well. It is encouraged to add a prompt for extra context.

Author: Done, updated the second paragraph of all the sections here to active

Requires polls or quizzes in previous pages to feed into the state of the element, and the state will be calculated according to either the quizzes (if thresholds are specified) or the polls (if only categories are specified) that the user answered.
Each category can specify an image, title and description that will be shown when the component selects that category for the user.

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/results-raw.png" layout="intrinsic" width="400" height="500">

[sourcecode:html]
<amp-story-interactive-results
    theme="dark"
    prompt-text="You are a"
    option-1-text="Dog" option-1-image="dog.png"
    option-1-description="You always have energy and love being with friends. Outdoors is your favorite place"
    option-2-text="Cat" option-2-image="cat.png"
    option-2-description="Cats are great animals for WFH">
</amp-story-interactive-results>
[/sourcecode]

## Attributes

The components have a shared API language for customizing their options. This makes it easier to port the configuration steps across all the interactive components.

### id (required for binary-poll, poll, quiz)

> Identifies the interactive component, and is used to compose the interactiveId that is sent to the backend. Should be unique for each component in the story.

Author: I linked `interactiveId` to the section below that explains it better. Not sure how to update it to active voice though... Would "...composes the interactiveId..." work? LMK

### endpoint (required for binary-poll, poll, quiz)

Complete URL of the backend, where votes for the interactive components are stored.

### theme (optional)

> Controls the color of the chips and text. Can be light (default), dark.

Author: You set the attribute to `light` or `dark`. Added backticks here and below to make it easier to understand that those are literal values. Still, the examples at the beginning have `theme="dark"` if someone wants to understand

### chip-style (optional for poll, quiz, results)

> Controls the style of the component. Can be flat (default), shadow, or transparent.

Author: Added backticks, check prev comment

Results components don't support shadow.

### prompt-text (optional)

Adds a prompt on top of the component, useful for writing the question to a quiz or poll. Also used to write a prompt to the results component before the category title (highly encouraged).

> ### prompt-size (optional for binary-poll, poll, quiz)

Author: It's not allowed in results since they don't allow prompts. I added a bit more to the explanation

> ### prompt-size (optional for binary-poll, poll, quiz)

Author: From the images on the documentation it's quite clear that the binary-poll, poll and quiz add a top bar that contains the prompt in a consistent manner. The results component doesn't have that top bar, and the prompt is added to the middle of the element, so it's not flexible with the size. This means this attribute only affects those elements listed above

> Controls the font-size of the prompt text. Can be small, medium (default), large. Large prompts will hold up to 3 lines of text, other sizes will hold up to 4 lines of text.

Author: Same thing as before, I added backticks to explain that you add those values to the attribute. I don't think the pixel sizes are necessary, as we provide many other places if people want to check the sizes. What's more important is that if there's a lot of text, then creators should set the size to small, and if there's very few words, they should set it to large (so that it looks good). It's a purely creative decision that the way people will get exposure to it is through storymaking tools as a dropdown for instance. Does that make sense or should we still add the pixel values?

### option-{1/2/3/4}-text (required)

> String that represents the option. On a results component, it determines the description for the category associated (highly encouraged). Binary polls require 2 options, but polls, quizzes and results can have 2-4 options.

Author: We have validation ensuring users need to define at least options 1 and 2, so the minimum is 2 options. I made sure to write in the documentation for tools that tools shouldn't allow users to add a number of options that is not supported.

### option-{1/2/3/4}-confetti (optional for binary-poll, poll, quiz)

> Emoji that is used in a confetti burst animation when the option is selected. When specified on an option, the confetti burst will be activated. On quizzes, only the correct option should have a confetti.

Author: We check for the content to be an emoji using regex, so any emoji is valid but no text or other things. If we ever add support for that we will update the docs here.

### option-{1/2/3/4}-results-category (optional for poll, required for results)

On the results component, it is used as the name of the category. On polls it links the options to the result with that name. The string has to match for the options to be linked.

> If the results component doesn't specify thresholds, the category strategy will be used: the category with more options selected in polls across the story will be shown.

Author: Changed the organization of the info

### option-{1/2/3/4}-results-threshold (optional for results)

On the results component, it is used as a lower boundary for the category when linked to quizzes. The component will calculate the score as a percentage of questions answered correctly (between 0 and 100), and it will show the category that has the best lower threshold that matches the score. The best threshold is the highest one that is lower or equal to the score, or the lowest score if all thresholds are higher than the score. If a threshold is specified for any option, all other options also need a threshold.

> ## Aggregate data source

Author: This is actually not as related to the component as it is for informational purposes, so it's not part of the usage of the elements. The context here is that all the elements listed below require to be connected to a backend, but tools can make their own backends if they follow these API standards. Happy to discuss where we should put this information now that you have this extra context, or what header to use. I think "display selection percentage" is not the focus of this section as it is the storage of aggregate data. What do you suggest?

> ## Aggregate data source

Reviewer 2: Usage is meant to encompass information that is vital to successful use of the component. If all components require a backend, this is relevant to usage. 

With that, reference documentation is for developers building experiences with these components - tools and integrations are our secondary audience. 

I still feel this section needs to be moved and renamed. Maybe "Store and display user selections"

> ## Aggregate data source

Author: Ok

Response: {
"options": [
{"index": 0, "selected": false, "count": 0},
{"index": 1, "selected": false, "count": 5},
{"index": 2, "selected": false, "count": 7},
{"index": 3, "selected": false, "count": 2}
]
}
[/sourcecode]

[sourcecode:js]
// Posting the vote for an interactive. Client param is required.

POST_URL = "{endpoint}/{interactiveId}:vote?type={interactiveType}&client={ampId}"
POST_BODY = {'option_selected': 1}

Response: No response necessary
[/sourcecode]

Backends need to be specified on the necessary components (binary-poll, poll, quiz), and can be deployed by publishers, tools or others. Available free-to-use backends are:

- Google hosted: _coming soon_

[tip type="note"]
Before setting up a backend, consider if the already existing backends satisfy your requirements.
[/tip]

## Styling

If you want to see all the theming options in action, check out the [example story](/documentation/examples/components/amp-story-interactive-poll/story#page=title-themes)

### CSS Variables

All interactive components support similar styling, which is controlled through CSS variables and attributes.
CSS variables

> Some variables can be overridden through assigning a class to the component.

Author: Hehe for some reason one of the copy/pasting included the title twice. Fixed that and used your suggesting

> - `--interactive-accent-color`: The accent color of the component. If no prompt-background is specified, it will use that for the prompt background as well.

Reviewer 2: I still don't understand this sentence :)

> - `--interactive-accent-color`: The accent color of the component. If no prompt-background is specified, it will use that for the prompt background as well.

Author: I rewrote some of these CSS variables to make them easier to understand

> - `--interactive-prompt-text-color`: Color of the top text. Only used on selectable components if a

Author: Used that naming on the top as well, it refers to the elements that have options that can be selected. Quizzes, polls and binary polls can be selected, but results are not selected (they are inferred). Since "components" is not the right word here, I'll use elements

  prompt is specified, or on results components if there are thresholds (which will color the score).
- `--interactive-prompt-background`: Background of the top text. Only used on selectable components if a prompt is specified, or on results if there are thresholds but no image (which will color the score background). Can be a color (including transparent) or CSS gradient.
- `--interactive-prompt-alignment`: Alignment of the prompt. Only used in selectable components (not results). Will default to center if the component has the transparent style or if it's a binary poll, otherwise it will default to initial.

### Themes

The `theme` attribute controls the chip color and text color of the component. Can be `light` (default) or `dark`.
The `chip-style` attribute controls the style details of the component. Can be `flat` (default), `shadow` or `transparent`.

## Sizing

> The component follows the [container](https://amp.dev/documentation/guides-and-tutorials/learn/amp-html-layout/layouts_demonstrated/#container) model. The size can be changed by overriding the element's font-size, which by default will be set to `3*var(--story-page-vmin)`. The component has a `min-width: 14em` and `max-width: 25em`, unless it's a binary-poll or results, in which case it will have `max-width:18em`. This makes it occupy 75% of the width on portrait stories by default, but the width can be overridden with a CSS rule to any value in between the min and max. The height will depend on the font-size and number of lines on the prompt, so it cannot be specified.

Author: Sounds good, I updated the text a lot to match more what you wrote. WDYT now?

> The component follows the [container](https://amp.dev/documentation/guides-and-tutorials/learn/amp-html-layout/layouts_demonstrated/#container) model. The size can be changed by overriding the element's font-size, which by default will be set to `3*var(--story-page-vmin)`. The component has a `min-width: 14em` and `max-width: 25em`, unless it's a binary-poll or results, in which case it will have `max-width:18em`. This makes it occupy 75% of the width on portrait stories by default, but the width can be overridden with a CSS rule to any value in between the min and max. The height will depend on the font-size and number of lines on the prompt, so it cannot be specified.

Reviewer: I'm wondering how much we could simplify this block of text, by saying it acts just like a paragraph of text. No one knows what `3*var(--story-page-vmin)` is (I barely do to be honest).

What's the one information we're trying to communicate here? If it is that we need to use a `font-size` in `em` to size it, we should make it that simple, with examples.

### Creating pixel-perfect layouts using amp-story-grid-layer[aspect-ratio]

> While the component by default adapts to the screen size with the variable font-size, it doesn't stay perfectly consistent across screen sizes. It's possible to use the aspect-ratio layer in order to create layouts that will scale perfectly with different screen sizes, by setting the font-size in ems on the component.

Author: "the elements" is what should be here instead. All the elements in this document have an adaptive size

Binary polls only on post-selection will show different font-sizes for the options. They will hold a large font-size if all the options are either emojis or have short texts (length <= 3). Medium font-size will be used if at least one of the options are longer than 3 characters but both can be displayed in one line; and small font-size will be used if at least one option has two lines.

Polls will also adapt the font-size of the options depending on the content. If all the options fit in one line, they will use a large font-size. If any of the options require 2 lines, all the options will be reduced in size to accommodate the option text in the chip.

For a live demo, check out the [binary-poll](https://codepen.io/mszylkowski/pen/oNxogoV) and [poll](https://codepen.io/mszylkowski/pen/ZEWaBoZ) Codepens, and change the option texts (be sure to select "Answered" on the binary-poll demo to see the size change).

## Animations

> The designs provide a subtle animation that enhances the interactivity aspect of the component, and are also animated when an option is selected to show the aggregate data.

Author: The animations can be seen if users go to the example story, and the elements will show the animation when a user lands on a page. I'll add some more documentation on that

> To add entering animations, [amp-story animations](https://amp.dev/documentation/components/amp-story/?format=stories#animation-attributes) can be added to all components.

Author: Yes! The animations of the elements (in the first sentence) are animating the options individually, so it's an animation of the inner elements. Usually the options animate in cascade to emphasize the options. amp-story-animation provides element-wide animations such as fade-in, scale up, etc.

## Analytics

> The component comes with support for [amp-analytics](https://amp.dev/documentation/components/amp-analytics/) events out of the box with the event `story-interactive`, and will report when an option is selected:

Reviewer 2: The `amp-story-interactive` component elements support [`amp-analytics`](https://amp.dev/documentation/components/amp-analytics/). Report a selected option by adding the `story-interactive` event to your configuration:

- `storyInteractiveId`: the element id
- `storyInteractiveResponse`: the option selected
- `storyInteractiveType`: the enum InteractiveType

```html
<amp-analytics id="my-analytics">
  <script type="application/json">
    {
      "requests": {
        "interactive": "https://example.com/interactive?id=${storyInteractiveId}&response=${storyInteractiveResponse}"
      },
      "triggers": {
        "interactiveSelected": {
          "on": "story-interactive",
          "request": "interactive"
        }
      }
    }
  </script>
</amp-analytics>
```

## Validation

See validation rules in [amp-story-interactive validator](https://github.com/ampproject/amphtml/blob/master/extensions/amp-story-interactive/validator-amp-story-interactive.protoascii).

## References

> Check this [Codepen collection](https://codepen.io/collection/DEGRLE) to play with the components and styles.

Author: Sounds good. If it makes sense, I moved the tip to the styling section, since the tip is mostly related to the styles

    option-4-text="Winter" option-4-confetti="☃️">
</amp-story-interactive-poll>
[/sourcecode]

### amp-story-interactive-quiz

The `amp-story-interactive-quiz` element provides a guessing experience with 2-4 options, one of which is correct. It displays the voting percentages after the user makes a selection. The user selection is green if correct and red if incorrect.

Display different categories based on percentage of correct user answers by pairing `amp-story-interactive-quiz` with `amp-story-interactive-results`. Add a prompt for extra context.

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/quiz-raw.png" layout="intrinsic" width="400" height="450">

[sourcecode:html]
<amp-story-interactive-quiz
    id="arts-quiz"
    endpoint="https://endpoint.com/v1/interactives"
    prompt-text='Who was the artist that created the famous painting "The Last Supper"?'
    option-1-text="Michelangelo"
    option-2-text="Leonardo da Vinci" option-2-correct option-2-confetti="🎨"
    option-3-text="Rahael"
    option-4-text="Sandro Boticelli">
</amp-story-interactive-quiz>
[/sourcecode]

### amp-story-interactive-results

> The `amp-story-interactive-result` element displays a customized state defined by the user's selection from previous polls or quizzes. This element requires use of a poll or quiz from previous pages to calculate an answer-based state. Each result state uses a category that may include an image, title and description to display to the user.

Author: Added more info on that

Complete URL of backend. Stores interactive element votes.

### theme (optional)

Controls the color of the chips and text. Can be `light` (default), `dark`.

### chip-style (optional for poll, quiz, results)

Controls the style of the component. Can be `flat` (default), `shadow`, or `transparent`.
Results and binary-poll elements don't support shadow.

### prompt-text (optional)

Adds a prompt to the top of the component. Use `prompt-text` to write the poll/quiz question, or as a prefix to the category in the `amp-story-interactive-result` element.

### prompt-size (optional for binary-poll, poll, quiz)

Controls the `font-size` of the prompt text, so it only applies for elements that have prompts. Can be `small` (), `medium` (default), `large`. Large prompts will hold up to 3 lines of text, other sizes will hold up to 4 lines of text.

### option-{1/2/3/4}-text (required)

String that represents a numbered option. Binary polls require 2 options. Polls and quizzes may include between 2 and 4 options.

The `amp-story-interactive-result` element uses this string value as category description.

### option-{1/2/3/4}-confetti (optional for binary-poll, poll, quiz)

> Emoji that bursts in an explosion animation when an option is selected. On quizzes, only the correct option should have a confetti.

Author: All emojis are supported, we need to properly validate that the input is just emojis, but we don't limit the emojis used here

### option-{1/2/3/4}-results-category (optional for poll, required for results)

On the results component, this attribute represents the name of the category, shown in larger text after the `prompt-text` and before the category description. If the results element doesn't specify thresholds, the category strategy will be used: the category with more options selected in polls across the story will be shown.

On polls it links the options to the result with that name as mentioned above. The string has to match perfectly for the options to be linked.

> ### option-{1/2/3/4}-results-threshold (optional for results)

Author: Added an example on thresholds and categories linked to the state.

On the results element, it determines the lower boundary for the category when linked to quizzes. The component will calculate the score as a percentage of questions answered correctly (between 0 and 100), and it will show the category that has the best lower threshold that matches the score. The best threshold is the highest one that is lower or equal to the score, or the lowest score if all thresholds are higher than the score. If a threshold is present for any option, all other options also need a threshold.

## Aggregate data source

> All selectable interactive elements (binary-poll, poll, quiz) show the percentage of users that selected each option. The backend specified with the `endpoint` attribute will store the aggregate data for the interaction following the API described below.

Reviewer 2: Please list the element names in full.

To fetch the data for an interactive element, the necessary fields are:

> - <div id="interactiveId"></div>`interactiveId`: the `base64encode(CANONICAL_URL) + "+" + element.id`

Reviewer 2: something is weird with the formatting here - maybe place a space between the closing div tag and the backtick?

Author: Sure, I want to link another part of this doc to here and seems to work. Will add a space

// If no client param, selected will always be false
// (can be used to display the results on a dashboard).

GET_URL = "{endpoint}/{interactiveId}?type={interactiveType}&client={ampId}"

Response: {
"options": [
{"index": 0, "selected": false, "count": 0},
{"index": 1, "selected": false, "count": 5},
{"index": 2, "selected": false, "count": 7},
{"index": 3, "selected": false, "count": 2}
]
}
[/sourcecode]

[sourcecode:js]
// Posting the vote for an interactive. Client param is required.

POST_URL = "{endpoint}/{interactiveId}:vote?type={interactiveType}&client={ampId}"
POST_BODY = {'option_selected': 1}

Response: No response necessary
[/sourcecode]

Backends need to be specified on the necessary components (binary-poll, poll, quiz), and can be deployed by publishers, tools or others. Available free-to-use backends are:

- Google hosted: _coming soon_

[tip type="note"]
Before setting up a backend, consider if the already existing backends satisfy your requirements.
[/tip]

## Styling

View all theming options in action in the [example story](https://amp.dev/documentation/examples/components/amp-story-interactive-poll/story#page=title-themes).

[tip type="read-on"]

> Check this [Codepen collection](https://codepen.io/collection/DEGRLE) to play with the components and styles.

Reviewer 2: ```suggestion
View and play with `amp-story-interactive` elements and styles in this [Codepen collection](https://codepen.io/collection/DEGRLE)!
```

Binary polls only on post-selection will show different font-sizes for the options. They will hold a large font-size if all the options are either emojis or have short texts (length <= 3). The element will use medium font-size if at least one of the options are longer than 3 characters but both can be displayed in one line; and small font-size if at least one option has two lines.

Polls will also adapt the font-size of the options depending on the content. If all the options fit in one line, they will use a large font-size. If any of the options require 2 lines, all the options will reduce in size to accommodate the option text in the chip.

For a live demo, check out the [binary-poll](https://codepen.io/mszylkowski/pen/oNxogoV) and [poll](https://codepen.io/mszylkowski/pen/ZEWaBoZ) Codepens, and change the option texts (be sure to select "Answered" on the binary-poll demo to see the size change).

> ## Animations
> 
> The designs provide a subtle animation on page-enter that enhances the interactivity aspect of the component. Transitioning animations also triggered when an option is selected to show the aggregate data.
> To add entering animations, [amp-story animations](https://amp.dev/documentation/components/amp-story/?format=stories#animation-attributes) `animate-in` attributes can be added to all elements.
> 

Reviewer 2: ```suggestion
### Animations

Enhance interactivity by adding an animation to the element when entering the page or when transitioning from option selection to data display. Pick from the [`amp-story` animations](https://amp.dev/documentation/components/amp-story/?format=stories#animation-attributes) and add the `animate-in` attribute to desired elements.

```

See the License for the specific language governing permissions and
limitations under the License.
-->

[tip type="important"]Change image sources to ampproject/amphtml/master!!![/tip]

The `amp-story-interactive` component provides a set of experiences, such as quizzes or polls, in [Web Stories](https://amp.dev/documentation/guides-and-tutorials/start/create_successful_stories/?format=stories).

<div layout="container" width="3" height="2">
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/quiz-art.png" layout="responsive" width="200" height="350"/></div>
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/animal-poll.png" layout="responsive" width="200" height="350"/></div>
<div style="width:32%;display:inline-block"><amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/results-art.png" layout="responsive" width="200" height="350"/></div>
</div>

## Usage

> The amp-story-interactive component encompasses a set of interactive experiences. Specify an interactive experience by defining one of the elements below. For best results, only use one element per [`amp-story-page`](https://amp.dev/documentation/components/amp-story-page/?format=stories).

Reviewer 2: ```suggestion
The `amp-story-interactive` component encompasses a set of interactive experiences. Specify an interactive experience by defining one of the elements below. For best results, only use one element per [`amp-story-page`](https://amp.dev/documentation/components/amp-story-page/?format=stories).
```

Each category specifies its content `option-{1/2/3/4}-{text/image/results-category}` attributes, where `-results-category` refers to the name of the category, `-image` refers to the image that will be displayed if that category is selected, and `-text` refers to the description.

Results can feed its state from quizzes if all categories also specify `option-{1/2/3/4}-results-threshold`. If no categories have thresholds, the state will count the `option-{1/2/3/4}-results-category` from options selected in polls.

<amp-img src="https://github.com/mszylkowski/amphtml/raw/interactive_docs/extensions/amp-story-interactive/img/results-raw.png" layout="intrinsic" width="400" height="500">

[sourcecode:html]
<amp-story-interactive-results
    theme="dark"
    prompt-text="You are a"
    option-1-results-category="Dog" option-1-image="dog.png"
    option-1-text="You always have energy and love being with friends. Outdoors is your favorite place"
    option-2-results-category="Cat" option-2-image="cat.png"
    option-2-text="Cats are great animals for WFH">
</amp-story-interactive-results>
[/sourcecode]

### Store and display user selection

> All selectable interactive elements (amp-story-interactive-binary-poll, amp-story-interactive-poll, amp-story-interactive-quiz) show the percentage of users that selected each option. The backend specified with the `endpoint` attribute will store the aggregate data for the interaction following the API described below.

Reviewer 2: ```suggestion
All selectable interactive elements (`amp-story-interactive-binary-poll`, `amp-story-interactive-poll`, `amp-story-interactive-quiz`) show the percentage of users that selected each option. The backend specified with the `endpoint` attribute will store the aggregate data for the interaction following the API described below.
```
